### PR TITLE
Multiprocessing executor

### PR DESCRIPTION
**Pull request by schrockn**
Here's the first show at the multiprocessing executor.

Next steps:

1) Actually exploit parallelism with a configurable number of parallel processes
2) Actually stream real events that represent execution plan start and and stop events rather than the hardcoded 'DONE' string
3) Perhaps use the filesystem as the intermediate values instead of holding everything in memory.

### CODE REVIEW

File: python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            success_data=result['success_data'],
+            failure_data=result['failure_data'],
+            tags=result['tags'],
+        )
+
+    # Do something reasonable on total process failure
+    process.join()
+
+
+def _all_inputs_covered(step, results):
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py]: isn't this identical to `simple_engine._all_inputs_covered`

File: python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    for step_input in step.step_inputs:
+        if step_input.prev_output_handle not in results:
+            return False
+    return True
+
+
+def _create_input_values(step, prev_level_results):
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py]: ditto -- these are common execution engine utils

File: python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    input_values = {}
+    for step_input in step.step_inputs:
+        prev_output_handle = step_input.prev_output_handle
+        input_value = prev_level_results[prev_output_handle].success_data.value
+        input_values[step_input.name] = input_value
+    return input_values
+
+
+class MultiprocessExecutorConfig:
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py]: can you elaborate on this -- particularly, is this where we're expecting to do things like set parallelism, etc

Comment (schrockn) [reply python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py]: yeah i'm going to work on that next. degrees of parallelism would be a property of this config object

File: python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py (hunk 1 above)
Comment (schrockn) [reply python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py]: yup it is the same. i expect these to diverge so i started with a copy paste job. I think we'll see this change substantially in subsequent work so i'd rather not consolidate the two code paths quite yet

File: python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+        yield ExecutionStepEvent(
+            event_type=event_type,
+            step=step,
+            success_data=result['success_data'],
+            failure_data=result['failure_data'],
+            tags=result['tags'],
+        )
+
+    # Do something reasonable on total process failure
```
[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/core/execution_plan/multiprocessing_engine.py]: This comment is in reference to a test case something like:

```py
@lambda_solid(
    inputs=[InputDefinition('sum_df', dagster_pd.DataFrame)],
    output=OutputDefinition(dagster_pd.DataFrame),
)
def crashy_solid(sum_df):  # pylint: disable=W0613
    os._exit(1)  # pylint: disable=W0212
```

which we use in the pipeline_execution_manager tests

File: python_modules/dagster/dagster/core/execution.py [hunk 5]
```diff
@@ -584,6 +602,25 @@ def __init__(self, pipeline, errors, config_value, *args, **kwargs):
         super(PipelineConfigEvaluationError, self).__init__(error_msg, *args, **kwargs)
 
 
+def invoke_executor_on_plan(pipeline_context, execution_plan, throw_on_user_error, executor_config):
+    if executor_config is None:
+        executor_config = InProcessExecutorConfig()
+
+    if isinstance(executor_config, InProcessExecutorConfig):
```
[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/core/execution.py]: I did that initially but for lack of a better term it just didn't feel right. There are a couple things going on here. One is that some executors require not just the pipeline object but enough information to reconstruct the pipeline. One implementation of that is just having the pipeline object directly, another would be the argless function, another would be the "RepositoryTargetInfo" stuff that dagit uses. In some ways that is separate from the executor.

I guess I just wanted a clean separation between external API (i.e. the user only needs to know about raw data in a config) versus having to look at a class that has a bunch of behavior. This felt like a case where a more pure functional approach seemed better.

To summarize: I don't have particularly cogent reasoning as to structuring this way, but my instincts told me to do this, so I did it. Not tied to it at all.
